### ISSUE TITLE
Development installation fails

### ISSUE THREAD

**Issue by johannesjh** [johannesjh/fava-gtk#2]
EDIT: This issue was migrated from where it was originally created at https://github.com/johannesjh/fava-desktop/issues/2. 
@yagebu wrote:

I tried installing it the development way, the first error was the following (I have poetry `1.0.10` installed):

```
The lock file might not be compatible with the current version of Poetry.
Upgrade Poetry to ensure the lock file is read properly or, alternatively, regenerate the lock file with the `poetry lock` command.
Installing dependencies from lock file

[NonExistentKey]
'Key "reference" does not exist.'
```

So I ran `poetry lock` and tried again. This started installing things but errored when trying to install and failing to build `pygobject==3.38.0`, which I already have installed system-wide.

**Comment by johannesjh** [comment]
EDIT: This comment was migrated from where it was originally created at https://github.com/johannesjh/fava-desktop/issues/2. 
@yagebu wrote:

I actually managed to get it running by just installing it with `pip install .`, however this will still forcibly install the vendored Fava. Is there a way to tell poetry to use system-wide packages?

**Comment by johannesjh** [comment]
well, the error message recommends to switch to a newer version of poetry. 

on my system:

**Comment by yagebu** [comment]
> Regarding the poetry version: well, the error message recommends to switch to a newer version of poetry. on my system, I currently have poetry version 1.1.4, freshly updated using pip.

I'm on Arch Linux so I expected to have a rather recent version installed, but apparently poetry is still a bit too niche to get updated quickly.

> What happens if you remove pygobject from pyproject.toml and then add it again using poetry add pygobject, does this work on your machine?

That fails again at trying to pip install pygobject. The pygobject packaging guide (https://pygobject.readthedocs.io/en/latest/packagingguide.html) suggests that using pip should work, but the linked packages for Arch and MingW use meson and I can't tell for Debian.

> The whole purpose of tools like poetry, pipenv, and virtualenv is to avoid messing with system-wide packages.

I know. But having access to system-wide installations is something different (and supported by virtualenv and pipenv at least, see, e.g., https://pipenv.pypa.io/en/latest/advanced/#working-with-platform-provided-python-components). It seems to me that for a dependency like pygobject, the poetry/pipenv approach doesn't really work. On my system, I have pygobject 3.38.0 installed and pip-building the exact same version fails. So I'd expect this to work even less on systems that have non-matching versions of GTK et al and the attempted version of pygobject.

I think it would be better to treat pygobject as a "system dependency". Currently, for installing fava-gtk, one needs to build and install pygobject which requires glib, libgirepository, libffi, and pkg-config. With pygobject, that would be reduced to one.

**Comment by johannesjh** [comment]
Some more research on this topic: PyGobject documents two options of installing / using it, see https://pygobject.readthedocs.io/en/latest/getting_started.html#ubuntu-getting-started

- option (1) Installing a system-provided PyGObject. On ubuntu, this means to install all packages through apt (no pip packages). 
- option (2) Installing from PyPI with pip. On ubuntu, this requires to first install development versions of python gi packages using apt, as well as build tools, so that python packages can then be installed through pip.

fava-gtk currently depends on PyGObject through a dependency declared in pyproject.toml... this probably corresponds to option (2), and this probably explains the problems you are experiencing. In other words: If you follow PyGObject's instructions "Installing from PyPI with pip", you should be able to install fava-gtk with no issues. The question remains if we should be using system packages instead.

To be honest I don't know how to proceed. 

One idea: Maybe poetry's [environment markers](https://python-poetry.org/docs/dependency-specification/#using-environment-markers) can be used to only conditionally declare the dependency on fava. But what should be the condition?

**Comment by johannesjh** [comment]
thank you for pull request https://github.com/johannesjh/fava-gtk/pull/3 and for releasing fava 1.18

this does not yet solve the installation problems that you have been experiencing, does it?

**Comment by yagebu** [comment]
> this does not yet solve the installation problems that you have been experiencing, does it?

Nope.

> I think the link to https://pipenv.pypa.io/en/latest/advanced/#working-with-platform-provided-python-components points in the right direction.

I think so too. However I didn't find a corresponding option for poetry, maybe that's not possible with poetry.

Of the 2 options you describe, I think option (1) is the way to go. Currently, you use poetry to a) manage a dev environment and b) to build the package. Annoyingly, `pip install -e` requires a setup.py, so that won't work with b). So one possible fix would be to switch to standard setuptools for the building of the package, then one could do a development install without being forced to use poetry.

**Comment by johannesjh** [comment]
I would not mind switching to pipenv if that solves the problem. Such a solution has also been recommended in this stackoverflow answer https://stackoverflow.com/a/64470393 but it looks a bit hacky to me:

```bash
pipenv --site-packages  # creates an isolated virtualenv, but breaks the isolation using the site-packages option
PIP_IGNORE_INSTALLED=1 pipenv install  # no clean way of telling pipenv where to use system packages vs virtualenv packages
```

It is unclear to me how we are supposed to tell pipenv that we want to use PyGObject from the system-wide site-packages, but that we want to version-lock and install all other dependencies into the virtualenv.

Besides pipenv, the experimental [PEP582](https://www.python.org/dev/peps/pep-0582/) looks like a simple solution. It installs version-locked dependencies into a local, .gitignore'd folder that is called `__pypackages__` by convention. Python preferrably load packages from that folder, but otherwise falls back to system-wide packages... which is exactly what we want. Not sure which package manager would be a good choice... the PEP582 ecosystem looks rather small. Maybe [PDM](https://github.com/frostming/pdm)

**Comment by yagebu** [comment]
> It is unclear to me how we are supposed to tell pipenv that we want to use PyGObject from the system-wide site-packages, but that we want to version-lock and install all other dependencies into the virtualenv.

Yep, that would be ideal, but I also can't tell if that's possible.

Didn't know about PEP582, that looks cool :)

**Comment by johannesjh** [comment]
fava-gtk has three groups of requirements:

* regular python dependencies: packages that can easily be installed through pip (e.g., fava, beancount)
* system dependencies: system packages that are difficult (or impossible) to install purely through pip (e.g., PyGObject)
* development dependencies: in our case, these can be installed through pip (e.g., pre-commit, pytest)

Unfortunately, tools like pipenv and poetry only support two groups of requirements, namely default and development requirements. The [pipfile readme](https://github.com/pypa/pipfile/blob/master/README.rst) explains this quite well:

> Existing requirements files tend to proliferate into multiple files - e.g. dev-requirements.txt, test-requirements.txt, etc. - but a Pipfile will allow seamlessly specifying groups of dependencies in one place. This will be surfaced as only two built-in groups (default & development). (see note below) [...] 
> Note
> Custom groups may be added in the future. Remember, it is easier to add features in the future than it is to remove them. The Composer community has been successful with only default and development as group options for many years. This model is being followed.

It seems to me that a simple, declarative solution does not exist. 
Multiple workarounds come to my mind.

A) One workaround is to ask developers/contributors (you) to please install PyGObject build dependencies, so that its installation using pip or poetry succeeds. This would allow to continue using poetry as-is, with all its convenience. Might not work on all machines, so an additional workaround would additionally make sense.

B) Another workaround would be to write a script that automates the creation of a virtual environment for local development purposes. This virtualenv would be created with access to the system's site_packages. The script would only install regular python dependencies into the virtualenv, no system dependencies. If we want the script to honor locked package versions, we would need to filter poetry's dependency tree. Not so easy.

C) Another workaround would be to use multiple requirements files, as in the old days. Different environments could install requirements from different requirements files. Versions could be frozen using pip-tools. 

D) Another workaround would be to specify the dependency on PyGObject as an extra. Usage of fava-gtk should fail gracefully if PyGObject does not exist.

**Comment by johannesjh** [comment]
I wrote a script that may serve as workaround in case pygobject cannot be installed using pip:
https://github.com/johannesjh/fava-gtk/blob/mkvenv/mkvenv.py
... the script creates a virtual environment and installs packages as specified in poetry's lock file. pygobject and its dependencies are not installed in the virtual environment, but used from the system-wide site packages.

Usage: A simple call to `./mkvenv.py` will create a virtual environment in a folder named `venv`.

Does this work for you?

**Comment by yagebu** [comment]
The `mkenv.py` seems to work insofar that it creates the venv. I'm not sure if that helps much though, since I still can't use poetry in any way. It seems to me that poetry and pipenv are just broken for a package like `fava-gtk` since they can't properly handle Python dependencies that should be treated as system dependencies...

**Comment by johannesjh** [comment]
Alright. I guess this leaves us with a custom-built solution, i.e., scripts that call pip and virtualenv. 

Note: I also looked at [pip-tools](https://pypi.org/project/pip-tools/) but I don't see its additional value and `pip-sync` does not seem to support virtualenvs with site-packages. I think it is therefore better to directly use pip and virtualenv.

> **`requirements-venv.in`**, for python requirements that are easily to install inside a pip virtual env:
> ```
> fava>=1.18
> # optionally some additional packages like smart-importer or beancount-share
> ```
> 
> ****`requirements-sys.in`**, for system requirements that are best used from the system-wide installation:
> ```
> PyGObject>=3.36
> ```
> 
> **`requirements-dev.in`**, for dev requirements:
> ```
> pytest>= 5.2
> pre-commit>=9.2.0
> pyclean~=2.0.0
> ```

A makefile like this could create the `requirement-*.txt` files with locked versions and update the virtual environment. Code for updating the venv is based on this article https://www.dinotools.de/en/2019/12/23/use-python-with-virtualenv-in-makefiles/

> **`Makefile`**
> ```make
> requirements-%.txt: requirements-%.in
> pip freeze $< --output-file $@
> 
> VENV_NAME?=venv
> PYTHON=${VENV_NAME}/bin/python
> 
> $(VENV_NAME)/bin/activate: requirements-venv.txt
> test -d $(VENV_NAME) || virtualenv -p python3 $(VENV_NAME)
> ${PYTHON} -m pip install -U pip
> ${PYTHON} -m pip install -r requirements-venv.txt
> ${PYTHON} -m pip install -r requirements-dev.txt
> touch $(VENV_NAME)/bin/activate
> ```

I currently don't have my development laptop, but I hope I can implement this in a few days time.

**Comment by johannesjh** [comment]
I moved away from poetry. Requirements are now declared in the requirements folder in text files. I use [pip-compile-multi] to lock package versions. I use a makefile target to create the virtual environment for developing purposes. This should also work on your machine. Specifically, `make venv` creates a virtual environment, installs all packages using their frozen versions, except for PyGObject which is is used from the sytem-wide site packages. Let me know if this works for you! - thanks!

**Comment by yagebu** [comment]
Mhh, this still seems to error on creating the venv - but an editable install now works which is more than enough for me :)

One tiny suggestion: Maybe split the generation of the requirements from the creation of the venv. That way one doesn't need pip-compile-multi installed when one is not interested in updating the requirements. Also, locking the requirements made some changes to the generated requirements files.

**Comment by johannesjh** [comment]
Thank you for your feedback! 

Regarding the creation of requirement .txt files: I removed the according makefile targets (not really needed), but I documented in the README how to upgrade dependencies.
* `pip-compile-multi` will upgrade dependencies, hence your observation that .txt files were changed
* `pip-compile-multi --no-upgrade` will re-generate the .txt files without upgrading dependencies.

Regarding the creation of the venv: I think I had a bug in the makefile. I fixed it in 7fd0ed67d3b973e6d694e40789a3909361c5ae8f, it should work now, hopefully :-) The system-wide installation of PyGObject will be used if it is available.
